Xamarin.Essentials/Screenshot/Screenshot.macos.cs

```diff
@@ -0,0 +1,56 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+using AppKit;
+using CoreGraphics;
+using Foundation;
+using ImageIO;
+using MobileCoreServices;
+
+namespace Xamarin.Essentials
+{
+    public static partial class Screenshot
+    {
+        internal static bool PlatformIsCaptureSupported => true;
+
+        static Task<ScreenshotResult> PlatformCaptureAsync()
+        {
+            var image = ApplicationServices.GetScreenshot();
+            var result = new ScreenshotResult(image);
+            return Task.FromResult(result);
+        }
+    }
+
+    public partial class ScreenshotResult
+    {
+        readonly CGImage cgImage;
```

Comment:
This will have to be disposed or risk being pushed on the finalization queue

Reply:
Sorry, but have no idea how to do that. `UiImage` is also not disposed in iOS:
https://github.com/xamarin/Essentials/blob/0dba844e8085f7cb36d62115b005620d541e0fbd/Xamarin.Essentials/Screenshot/Screenshot.ios.tvos.cs#L23

Reply:
That is what I realised as well, which is kind of a bummer. UIImage does not have to be disposed, CGImage has to be though. If you compare the docs [UIImage](https://docs.microsoft.com/en-us/dotnet/api/uikit.uiimage?view=xamarin-ios-sdk-12#methods) [CGImage](https://docs.microsoft.com/en-us/dotnet/api/coregraphics.cgimage?view=xamarin-ios-sdk-12#methods) you will see that while both expose a Dispose method for NSObject, CGImage does not have the same Dipose method as NSObject while UIImage does.

Reply:
Found a solution which I belive is ideal. You can remove the macos specific ScreenshotResult, convert the CIImage to UIImage via the constructor and alleviate the problems

Reply:
https://docs.microsoft.com/en-us/dotnet/api/uikit.uiimage.-ctor?view=xamarin-ios-sdk-12#UIKit_UIImage__ctor_CoreImage_CIImage_

Reply:
Is this possible in macOS SDK? I think not because I need a reference to UIKit in iOS?

Reply:
You are correct,you can not eliminate the struct but we can eliminate cgimage by moving to nsimage, which will follow the guidance to get away from native objects as soon as possible. NSImage also has a constructor accepting CGImage but also requires a CGSize trivially constructable by CGImage

Reply:
`CGImageDestination` not accept nsImage so we back again 😔

Reply:
How do you think about the added finalizer?

Reply:
Check out this gist: https://gist.github.com/Mrnikbobjeff/7fb533642a1b9bfeb7376a79d43da7dc

Reply:
And if you had to implement a finalizer be sure to check out [this](http://joeduffyblog.com/2005/04/08/dg-update-dispose-finalization-and-resource-management/) or take a look at my record audio pr recording class, it has the correct finalization + disposing pattern

Reply:
Oh nice. Great, thank you. Making PRs is the best way to learn 🦄 
I have committed the changes.

`CGImage.ScreenImage()` is not working correctly. I also added this as comment and in the PR description.

Reply:
I am sorry, I have missed disposing the actual stream in the custom stream class dispose method

Reply:
And what do you mean it does not work correctly?

Reply:
Oh my... time for weekend...
I added it, perhaps you have a look (again). Thank you!

`CGImage.ScreenImage()` have two problems:

**Not possible to catch entire screen**
![image](https://user-images.githubusercontent.com/2002916/95571256-6f776f80-0a28-11eb-95b0-d0fe737ea67c.png)

**Not working if the window is not on the main screen (empty)**
![image](https://user-images.githubusercontent.com/2002916/95571294-7c945e80-0a28-11eb-8e28-57a7f5973624.png)

Reply:
The only concern I have is app store releaseability, I do not know the policy for this for the mac app store. I am pretty sure that CIImage methods are allowed so I will follow up on that. Otherwise this looks great :)

Reply:
Next week I will try to commit Cuterdio to Mac App Store... So I can test.
Thanks for your help!

Reply:
https://github.com/xamarin/xamarin-macios/blob/caba0a66be1490b28f5fb38b17919c3eabebee56/src/CoreGraphics/CGImage.cs#L222

Reply:
/* Flags for CGWindowImageOption values.  These may be ORed together. */

enum {

  /* If `CGRectNull' is passed as the screen bounds, then then bounds
     computation includes window frame ornamentation, such as a shadow. */
  kCGWindowImageDefault = 0,

  /* If `CGRectNull' is passed as the screen bounds, then then bounds
     computation excludes window frame ornamentation, such as a shadow. */
  kCGWindowImageBoundsIgnoreFraming = (1 << 0),

  /* Force the created image to be opaque.  Empty areas are white */
  kCGWindowImageShouldBeOpaque = (1 << 1),

  /* Only draw the windows' shadows, not the windows themselves. */
  kCGWindowImageOnlyShadows = (1 << 2),

  /* Return the best image resolution. The screen size may be
     different than the returned image size. */
  kCGWindowImageBestResolution = (1 << 3),

  /* Return the nominal image resolution. The screen size
     equals the returned image size. */
  kCGWindowImageNominalResolution = (1 << 4)
};

/* Create an image containing a composite of the specified set of windows
   contained within a rectangular area. The set of windows is specified
   using options from `CGWindowListOption', along with an optional
   additional window ID.

   The windows list options are:

   --- kCGWindowListOptionAll, kCGWindowListOptionOnScreenOnly: Use all
   on-screen windows in this user session to construct the image. The
   parameter `windowID' should be `kCGNullWindowID'.

   --- kCGWindowListOptionOnScreenAboveWindow: Use all on-screen windows in
   this user session above the window specified by `windowID', ordered from
   front to back, to construct the image. To include the window specified by
   `windowID', add the flag `kCGWindowListOptionIncludingWindow'.

   --- kCGWindowListOptionOnScreenBelowWindow: Use all on-screen windows in
   this user session below the window specified by `windowID', ordered from
   front to back, to construct the image. To include the window specified by
   `windowID', add the flag `kCGWindowListOptionIncludingWindow'.

   --- kCGWindowListOptionIncludingWindow: Use only the window specified by
   `windowID' to construct the image.

   The parameter `screenBounds' specifies the rectangle in screen space
   (origin at the upper-left; y-value increasing downward). Setting
   `screenBounds' to `CGRectInfinite' will include all the windows on the
   entire desktop. Setting `screenBounds' to `CGRectNull' will use the
   bounding box of the specified windows as the screen space rectangle.

   The parameter `imageOptions' allows you to specify whether the window
   frame ornamentation, such as a shadow or similar effect, should be
   included or excluded in the bounds calculation when `CGRectNull' is
   specified for the window bounds.

   If no windows meet the specified criteria, or the windows can't be read,
   then a transparent black image will be returned.

   Any on-screen window with sharing type `kCGWindowSharingNone' will not
   be included in the image.

   This function returns NULL if the caller is not running within a Quartz
   GUI session or the window server is disabled. */